torax/sources/tests/electron_cyclotron_source.py

```diff
@@ -0,0 +1,164 @@
+        if not source_lib.is_source_builder(source_builder):
+            raise TypeError(f"{type(self)} has a bad _source_class_builder")
+        runtime_params = general_runtime_params.GeneralRuntimeParams()
+        source_models_builder = source_models_lib.SourceModelsBuilder(
+            {"foo": source_builder},
+        )
+        source_models = source_models_builder()
+        source = source_models.sources["foo"]
+        source_builder.runtime_params.mode = source.supported_modes[0]
+        self.assertIsInstance(source, source_lib.Source)
+        geo = geometry.build_circular_geometry()
+        dynamic_runtime_params_slice = (
+            runtime_params_slice.DynamicRuntimeParamsSliceProvider(
+                runtime_params=runtime_params,
+                sources=source_models_builder.runtime_params,
+                torax_mesh=geo.torax_mesh,
+            )(
+                t=runtime_params.numerics.t_initial,
+            )
+        )
+        core_profiles = core_profile_setters.initial_core_profiles(
+            dynamic_runtime_params_slice=dynamic_runtime_params_slice,
+            geo=geo,
+            source_models=source_models,
+        )
+        value = source.get_value(
+            dynamic_runtime_params_slice=dynamic_runtime_params_slice,
+            dynamic_source_runtime_params=dynamic_runtime_params_slice.sources["foo"],
+            geo=geo,
+            core_profiles=core_profiles,
+        )
+        # ECHCD provides TEMP_EL and PSI
+        chex.assert_rank(value, 2)
+
+    def test_invalid_source_types_raise_errors(self):
+        """Tests that using unsupported types raises an error."""
+        runtime_params = general_runtime_params.GeneralRuntimeParams()
+        geo = geometry.build_circular_geometry()
+        source_builder = self._source_class_builder()
+        source_models_builder = source_models_lib.SourceModelsBuilder(
+            {"foo": source_builder},
+        )
+        source_models = source_models_builder()
+        source = source_models.sources["foo"]
+        self.assertIsInstance(source, source_lib.Source)
+        dynamic_runtime_params_slice_provider = (
+            runtime_params_slice.DynamicRuntimeParamsSliceProvider(
+                runtime_params=runtime_params,
+                sources=source_models_builder.runtime_params,
+                torax_mesh=geo.torax_mesh,
+            )
+        )
+        dynamic_runtime_params_slice = dynamic_runtime_params_slice_provider(
+            t=runtime_params.numerics.t_initial,
+        )
+
+        core_profiles = core_profile_setters.initial_core_profiles(
+            dynamic_runtime_params_slice=dynamic_runtime_params_slice,
+            geo=geo,
+            source_models=source_models,
+        )
+
+        for unsupported_mode in self._unsupported_modes:
+            source_builder.runtime_params.mode = unsupported_mode
+            with self.subTest(unsupported_mode.name):
```

Comment:
I think you are missing a construction of a new `dynamic_runtime_params_slice` here (which will have the updated `unsupported_mode` inside it that should then throw an error 

```
source_builder.runtime_params.mode = unsupported_mode
dynamic_runtime_params_slice = (
          runtime_params_slice.DynamicRuntimeParamsSliceProvider(
              runtime_params=runtime_params,
              sources=source_models_builder.runtime_params,
              torax_mesh=geo.torax_mesh,
          )(
              t=runtime_params.numerics.t_initial,
          )
      )
with self.assertRaises(RuntimeError):
                    source.get_value(...
```

Reply:
Yep, that's the one! Thanks :)